### PR TITLE
Extend PSI feature selection to categorical variables

### LINKED ISSUE #655: PSI feature selection for categorical features
Currently, PSI feature selection as implemented in selection/drop_psi_features.py can only be applied to numerical features: it first converts these numerical features to categoricals via bucketing, then calculates PSI for the discretised numerical features. 

It would be useful to also be able to use DropHighPSIFeatures on categorical features. I believe the implementation is fairly straightforward.

I'd be happy to pick up this issue if it is of interest to others.

### PR DESCRIPTION
**Pull request by dlaprins**
closes #655 
closes #658

### CODE REVIEW

**Comment by ClaudioSalvatoreArcidiacono** [comment]
Hey @dlaprins ping me once this PR is ready for review!

File: feature_engine/selection/drop_psi_features.py [hunk 1]
```diff
@@ -7,42 +7,31 @@
...
-    _check_init_parameter_variables,
-)
-from feature_engine.variable_handling.variable_type_selection import (
-    find_or_check_numerical_variables,
-)
+from feature_engine.tags import _return_tags
+from feature_engine.variable_handling._init_parameter_checks import \
+    _check_init_parameter_variables
+from feature_engine.variable_handling.variable_type_selection import \
+    find_categorical_and_numerical_variables
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:27]: Could you format this with `black`?

```
black feature_engine/selection/drop_psi_features.py
```

You should first run `isort`, then `black`. Not the other way around

File: feature_engine/selection/drop_psi_features.py [hunk 2]
```diff
@@ -354,8 +345,12 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
         # If required exclude variables that are not in the input dataframe
         self._confirm_variables(X)
 
-        # find numerical variables or check those entered are present in the dataframe
-        self.variables_ = find_or_check_numerical_variables(X, self.variables_)
+        # # find variables or check those entered are present in the dataframe
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:348]:
```suggestion
        # find variables or check those entered are present in the dataframe
```

File: feature_engine/selection/drop_psi_features.py [hunk 3]
```diff
@@ -387,28 +382,27 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
...
-            test_discrete = bucketer.transform(test_df[[feature]].dropna())
+            # Bin the features if it is numerical and determine number of bins
+            if feature in self.num_variables_:
+                basis_discrete = bucketer.fit_transform(basis_df[[feature]].dropna())
+                test_discrete = bucketer.transform(test_df[[feature]].dropna())
+                n_bins = self.bins
+            elif feature in self.cat_variables_:
+                basis_discrete = basis_df[[feature]]
+                test_discrete = test_df[[feature]]
+                n_bins = X[[feature]].nunique()[0]
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:405]:
```suggestion
                n_bins = X[feature].nunique()
```

File: feature_engine/selection/drop_psi_features.py [hunk 4]
```diff
@@ -354,8 +345,12 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
         # If required exclude variables that are not in the input dataframe
         self._confirm_variables(X)
 
-        # find numerical variables or check those entered are present in the dataframe
-        self.variables_ = find_or_check_numerical_variables(X, self.variables_)
+        # # find variables or check those entered are present in the dataframe
+        (
+            self.cat_variables_,
+            self.num_variables_,
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:351]: Is it necessary for these two variables to be a new instance attribute? How about keeping them as simple variables?

File: feature_engine/selection/drop_psi_features.py [hunk 5]
```diff
@@ -577,10 +582,24 @@ def _get_cut_off_value(self, split_column):
...
+        """Threshold computation for chi-square test.
+
+        The threshold is given by
+            threshold = χ2(q,B−1) × (1/N + 1/M)
+        where
+        q = quantile of the distribution (or 1 - p-value),
+        B = number of bins/categories,
+        N = size of basis dataset,
+        M = size of test dataset.
+        See formula (5.2) from reference.
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:595]: I know this comes from the previous version of docstring, but is this reference actually valid? Does It refer to something in the class docstring? If it is referring to outside resources, like feature engine documentation, isn't it better to refer with a link?

File: feature_engine/selection/drop_psi_features.py [hunk 6]
```diff
@@ -577,10 +582,24 @@ def _get_cut_off_value(self, split_column):
...
 
-    def _calculate_auto_threshold(self, N, M, q=0.999):
-        # threshold = χ2(q,B−1) × (1/N + 1/M)
-        # where q - quantile (or 1 - p-value) B - number of bins,
-        # N - size of basis dataset, M - size of test dataset
-        # see formula (5.2) from reference
-        # taking q = 0.999 to get higher threshold
-        return stats.chi2.ppf(q, self.bins - 1) * (1.0 / N + 1.0 / M)
+    def _calculate_auto_threshold(self, N, M, bins):
+        """Threshold computation for chi-square test.
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:586]: could you also write docstring for the arguments?

File: feature_engine/selection/drop_psi_features.py [hunk 7]
```diff
@@ -251,6 +240,7 @@ def __init__(
         missing_values: str = "raise",
         variables: Variables = None,
         confirm_variables: bool = False,
+        p_value: float = 0.001,
```
Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:243]: Docstring for this new parameter is missing.

Comment (ClaudioSalvatoreArcidiacono) [inline feature_engine/selection/drop_psi_features.py:243]: I would suggest to add this parameter on a different order, for example after threshold, to show they are more closely related. Moreover usually feature engine selector ends with the parameters `variables` and `confirm_variables`.

Comment (ClaudioSalvatoreArcidiacono) [reply feature_engine/selection/drop_psi_features.py:243]: Not sure about this comment, as adding it at the end works better with backwards compatibility. I think the maintainers should have the final say on it.

File: feature_engine/selection/drop_psi_features.py (hunk 1 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:27]: Yep, I second this one. Not sure what happened here, but if there are no new imports, I'd prefer we don't change what was originally there. It is also easier to review.

Could we fix @dlaprins ?

File: feature_engine/selection/drop_psi_features.py (hunk 7 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:243]: Interesting discussion. But if we put it after threshold, users that are already familiar with the order of the params will be frustrated. Let's leave it for the time being.

File: feature_engine/selection/drop_psi_features.py [hunk 8]
```diff
@@ -335,6 +325,7 @@ def __init__(
         self.strategy = strategy
         self.min_pct_empty_bins = min_pct_empty_bins
         self.missing_values = missing_values
+        self.p_value = p_value
```
Comment (solegalli) [inline feature_engine/selection/drop_psi_features.py:328]: we'd need a check for this value, that is a float within so and so.

File: feature_engine/selection/drop_psi_features.py (hunk 2 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:348]: Please fix as per @ClaudioSalvatoreArcidiacono 's suggestion

File: feature_engine/selection/drop_psi_features.py (hunk 4 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:351]: I guess @dlaprins  is copying from what we do in another selector that works with both numerical and categorical.

We probably do not need to generate these new attributes. We could drop it. But if we do, then the user would not be able to easily find out which variables the transformer considered numerical and which ones categorical.

Simple controversial example: discrete variables encoded as object, will be considered categorical. Will the user be aware of this? 

I am not sure what is best here. I appreciate your input.

File: feature_engine/selection/drop_psi_features.py [hunk 9]
```diff
@@ -387,28 +382,27 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
...
+            if self.strategy == "equal_width":
+                bucketer = EqualWidthDiscretiser(bins=self.bins)
+            else:
+                bucketer = EqualFrequencyDiscretiser(q=self.bins)
 
         # Compute the PSI by looping over the features
         self.psi_values_ = {}
         self.features_to_drop_ = []
 
         for feature in self.variables_:
```
Comment (solegalli) [inline feature_engine/selection/drop_psi_features.py:396]: instead of looping in `self.variables_` and then `self.numerical_` we should just loop over numerical and then over categorical, and, like this, remove the nested looping.

Could we fix that please?

File: feature_engine/selection/drop_psi_features.py (hunk 3 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:405]: pls fix as per suggestion :)

File: feature_engine/selection/drop_psi_features.py (hunk 6 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:586]: Why are we changing the default parameters of the auto-threshold? We used to default q to 0.999 and now this is an obligatory parameter.

What's the logic?

fyi @glevv

File: feature_engine/selection/drop_psi_features.py (hunk 5 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:595]: This is a hidden method, so it won't be exposed to users. It is mostly for developers. If it is not clear which reference it refers to, then we should spell it out though.

File: feature_engine/selection/drop_psi_features.py [hunk 10]
```diff
@@ -464,6 +468,7 @@ def _observation_frequency_per_bin(self, basis, test):
                 how="outer",
             )
             .fillna(self.min_pct_empty_bins)
+            .replace(to_replace=0, value=self.min_pct_empty_bins)
```
Comment (solegalli) [inline feature_engine/selection/drop_psi_features.py:471]: why do we add this line?

File: feature_engine/selection/drop_psi_features.py (hunk 5 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:595]: Ok, I see that p_value comes here, that's why we change the logic of the auto-threshold. We need to update the dosctrings.

@glevv could you review this bit please?

**Review by solegalli — COMMENTED** [review]
Hey @dlaprins and @ClaudioSalvatoreArcidiacono 

Thank you so much for the active discussions and code changes!.

I added my comments at the back of @ClaudioSalvatoreArcidiacono 's

@dlaprins will you be able to update the code implementation?

Thanks a lot!

File: feature_engine/selection/drop_psi_features.py (hunk 4 above)
Comment (ClaudioSalvatoreArcidiacono) [reply feature_engine/selection/drop_psi_features.py:351]: I see now that saving categorical and numerical variables as instance attributes have been done in [_prediction/base_predictor.py](https://github.com/feature-engine/feature_engine/tree/main/feature_engine/_prediction#L12) indeed. It can make sense to save them as instance attributes for that purpose, but in that case I would comment that documentation for the new instance attributes is missing.

On the other hand, in the [information value selector](https://github.com/feature-engine/feature_engine/blob/main/feature_engine/selection/information_value.py#L217), those are not saved.

I personally think that both approaches can be correct.

File: feature_engine/selection/drop_psi_features.py [hunk 11]
```diff
@@ -387,28 +382,27 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
...
-        else:
-            threshold = self.threshold
-
-        # set up the discretizer
-        if self.strategy == "equal_width":
-            bucketer = EqualWidthDiscretiser(bins=self.bins)
-        else:
-            bucketer = EqualFrequencyDiscretiser(q=self.bins)
+        # set up the discretizer for numerical variables
+        if len(self.num_variables_) > 0:
```
[the remark this answers is not in the document]
Comment (ClaudioSalvatoreArcidiacono) [reply feature_engine/selection/drop_psi_features.py:386]: Using PSI if there are only categorical features is perfectly fine and correct, so in my opinion, I would not raise an error or a warning in case there are no numerical values.

File: feature_engine/selection/drop_psi_features.py (hunk 4 above)
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:351]: Let's not save them. @dlaprins could we update the code please?

File: feature_engine/selection/drop_psi_features.py (hunk 1 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:27]: Thanks for letting me know,  I was unaware of the correct ordering.

File: feature_engine/selection/drop_psi_features.py (hunk 4 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:351]: Will do

File: feature_engine/selection/drop_psi_features.py (hunk 5 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:595]: The reference is to reference [1] in the docstring of the class (i.e., Yurdakul (2018, https://scholarworks.wmich.edu/dissertations/3208/). I have left it as is, let me know if you want to see the reference spelled out more explicitly

File: feature_engine/selection/drop_psi_features.py (hunk 10 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:471]: If a categorical variable is of type 'object', and the split misses a category, the dataframe will have a NaN which is then solved by .fillna. However, in case the categorical variable is of type 'category', and the split misses a category, it will be correctly put to 0: thus it won't get filled and the PSI becomes infinite. 

This line ensures that empty buckets have a minimal value such that the PSI doesn't explode, similar to how it functions for 'object' type categoricals.

File: feature_engine/selection/drop_psi_features.py (hunk 6 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:607]: Wrote some form of docstring for the arguments, but not sure if it's in the format desired. Please let me know if you want it to be something else. 

The logic behind changing the arguments is that the auto-threshold is called on by the class, but the p-value (or quantile, as previously used) could not be defined by the user. I would deem this to be a necessary option in order to make use of the auto-threshold (which I think is very useful in and of itself). 

The bins argument needs to be added for auto-threshold to be compatible with categorical features. For details, please see my discussion with Claudio in the Issues thread.

**Comment by dlaprins** [comment]
Thank you both @ClaudioSalvatoreArcidiacono @solegalli  for your helpful suggestions and directions. I have tried to fix all the shortcomings you pointed out in the comments in my commit. Please let me know which issues remain unresolved and I'll solve those as well.

**Review by solegalli — COMMENTED** [review]
Hey @dlaprins 

Thanks for the code changes. I went over the code and discussion, here are some more thoughts:

- [ ] we are now not capturing numerical and categorical variables in new attributes, that's great. Later on in the logic we are still looping over self.num_variables_ even though it does not exist any more. Could you please fix?
- [ ] It is not necessary to loop over self.variables_ we can remove the nested looping. 
- [ ] n_bins is being reassigned for each feature, this is not necessary for numerical, only for categorical. Could we update?
- [ ] the threshold is calculated once for all variables, I think the idea is to use the same threshold for numerical, and a changing threshold for categorical. Could we reflect this in the code?
- [ ] The question still remains: should we expose p_values?
- [ ] it would be great if we could add the threshold calculation formula in the init docstrings as well.
- [ ] @glevv mentioned that the calculation would not work for this example: current implementation will give erroneous results if you have some differences in categories in the same columns (cat_1_train have categories 'A', 'B' and 'C', and cat_1_test have categories 'A', 'B' and 'D'). Could we add that as a test and make sure it works somehow? or raises an exception to alert users on what to do?

Thank you for the great work!

File: feature_engine/selection/drop_psi_features.py [hunk 12]
```diff
@@ -419,6 +434,16 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
             self.psi_values_[feature] = np.sum(
                 (test_distrib - basis_distrib) * np.log(test_distrib / basis_distrib)
             )
+
+            # Determine the appropriate threshold
+            if self.threshold == "auto":
```
[the remark this answers is not in the document]
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py:447]: Hi @solegalli,

Let me try to address all the comments (including the other ones you left) in this one reply: 

> Hey @dlaprins
> 
> Thanks for the code changes. I went over the code and discussion, here are some more thoughts:
> 
>     * [ ]  we are now not capturing numerical and categorical variables in new attributes, that's great. Later on in the logic we are still looping over self.num_variables_ even though it does not exist any more. Could you please fix?

An oversight indeed.

>     * [ ]  It is not necessary to loop over self.variables_ we can remove the nested looping.

I'm a bit confused here. What do you mean by nested looping? As it stands, I believe there is a single for-loop over self.variables and only an if-statement inside of that for-loop, no other for-loops (for what it's worth: the for-loop over self.variables is necessary due to the dropna(), I tried vectorizing the for-loop before I noticed this and got wrong results). 

 >     * [ ]  n_bins is being reassigned for each feature, this is not necessary for numerical, only for categorical. Could we update?

The only way I see to update this is to use 2 loops, one over categoricals and one over numericals. Then n_bins doesn't have to be checked for each feature in the numerical loop. Is that what you have in mind? 

>     * [ ]  the threshold is calculated once for all variables, I think the idea is to use the same threshold for numerical, and a changing threshold for categorical. Could we reflect this in the code?

The threshold calculation is inside the loop over self.variables. Similar to n_bins, we can have a single numerical threshold outside of a for-loop over numericals, while for categoricals the threshold has to be calculated inside of the for-loop. 

>     * [ ]  The question still remains: should we expose p_values?

I believe we should. As a user I would find it rather unpleasant if I cannot set my own p-value threshold for a statistical test. It's fine if a user does not know what p_value they should set, as there's a default value for the p-value that they can use. 

>     * [ ]  it would be great if we could add the threshold calculation formula in the init docstrings as well.

 Will do

>     * [ ]  @glevv mentioned that the calculation would not work for this example: current implementation will give erroneous results if you have some differences in categories in the same columns (cat_1_train have categories 'A', 'B' and 'C', and cat_1_test have categories 'A', 'B' and 'D'). Could we add that as a test and make sure it works somehow? or raises an exception to alert users on what to do?

Is this really the case? I believe this is tested for by Claudio's test (see line 35, 85, 100 of test_drop_high_psi_features.py): "drift_cat_1" is exactly such that basis_test contains only "A" while test_df contains only "B". This is also why I got errors before modifying _observation_frequency_per_bin() when I changed the "drift_cat_1" feature in the test to be a categorical feature. Please correct me if I'm wrong here. 

> 
> Thank you for the great work!

Thank you for the suggestions on how to improve the code. 

I have left the loop over self.variables as it is for now, but if you would prefer to have 2 loops instead (or if I'm just not understanding you properly), let me know and I'll correct it.

File: feature_engine/selection/drop_psi_features.py (hunk 5 above)
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py]: I fixed the reference as per your suggestion. 

It is not necessary for the user to set the p-value, but it is optional, so in that sense a user who just wants to run it without thinking about parameters can still do so with the auto-setting. I am of the opinion that it would be a very useful feature to be able to set your own p-value, especially if one has to justify such values to stakeholders. The functionality of `auto' threshold versus float threshold are not identical, as the former varies per number of categories whereas the latter is identical for all features.

File: feature_engine/selection/drop_psi_features.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (glevv) [reply feature_engine/selection/drop_psi_features.py]: Not sure about using PSI on categorical features is "perfectly fine". Chi2 is a better way of doing it.

File: feature_engine/selection/drop_psi_features.py [hunk 13]
```diff
@@ -387,28 +403,27 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None):
...
         for feature in self.variables_:
-            # Discretize the features.
-
-            basis_discrete = bucketer.fit_transform(basis_df[[feature]].dropna())
-            test_discrete = bucketer.transform(test_df[[feature]].dropna())
+            # Bin the features if it is numerical and determine number of bins
+            if feature in self.num_variables_:
+                basis_discrete = bucketer.fit_transform(basis_df[[feature]].dropna())
+                test_discrete = bucketer.transform(test_df[[feature]].dropna())
+                n_bins = self.bins
```
[the remark this answers is not in the document]
Comment (glevv) [reply feature_engine/selection/drop_psi_features.py:430]: I agree, if psi selection is extended to categorical features, then we need to use list of auto thresholds depending on number of bins in these features.

File: feature_engine/selection/drop_psi_features.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dlaprins) [reply feature_engine/selection/drop_psi_features.py]: What's the reason for considering chi2 to be better than PSI?

[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py]: chi2 is very well studied and specifically designed to compare the frequency of categorical distributions (here more info: https://uk.sagepub.com/en-gb/eur/discovering-statistics-using-r/book236067). I think at some point I also read the chi2 article, but long forgotten.

I am also not sure I'd go and say "perfectly fine" to use PSI instead of chi2. Is this statement just based on that thesis? How well accepted is it in the community?

In my opinion, what happens is that if you make something available out of the box, like we are doing here, then less experienced users may be mislead to think that this is a proper way of assessing categorical variables. And I think, they should use chi2, based on my understanding of the consensus.

The thesis says that the PSI behaves like chi2, OK, it doesn't change the consensus the way I see it.

A follow up question could be, why does one use PSI instead of using other tests to compare continuous distributions, like the KolmogorovSmirnov test. I personally do not have the answer to that question, I haven't used the PSI much. So on that line, if we do discretize continuous variables to use the PSI, then why not apply it on categoricals as well, it seems like less work.

So in summary, what we need to decide is if we should we raise a warning if there are no numerical variables. And we should probably expand the user documentation with some discussion on chi2 vs PSI for categorical variables.

File: feature_engine/selection/drop_psi_features.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py:447]: Hi @dlaprins yes, that was what I meant, 2 for loops instead of the for and the ifs. Sorry I wasn't clear.

Thank you for answering all the points!

File: feature_engine/selection/drop_psi_features.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ClaudioSalvatoreArcidiacono) [reply feature_engine/selection/drop_psi_features.py]: I think my comment had been misinterpreted. I did not want to say that it is perfectly fine to use one metric over the other. What I said is that it is perfectly fine to use PSI to measure feature drift over a dataset of categorical variables only and thus it is not needed to raise a warning in this situation.

**Comment by solegalli** [comment]
FYI @dlaprins 

Please rebase main, we just merged #660 :)

File: feature_engine/selection/drop_psi_features.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/selection/drop_psi_features.py]: Thank you!

**Comment by dlaprins** [comment]
> FYI @dlaprins
> 
> Please rebase main, we just merged #660 :)

Fixed, as well as split up the loop over variables to 2 loops over cat variables and num variables separately as requested.
